extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+thanks to the no-double-spending rule. Condition \verb|con2| ensures
+that only a single token of the currency is created, effectively
+making it non-fungible.
+
+It is of course possible to produce multiple script outputs to the
+same address of \verb|nftValidator nft| but only one of them can be spent,
+because the referenced tx output can only be spent once.
+
+Note that creating an NFT with this contract requires three
+transactions: One for the nftBootstrapTxOut transaction output, one
+for producing an output to the NFT address, and one that forges the
+token, consuming the output from the NFT address and the
+\verb|nftBootstrapTxOut|. Note that the bootstrap transaction output can be
+any unspent transaction output owned by us, in particular it can be
+the same output that pays the fees for the forging transaction.
+
+\subsubsection{Another example?}
+  
+\section{XUTXO-3: Non-fungible tokens}
+\label{sec:xutxo-3}
+
+We have shown above how NFTs can be implemented in the XUTXO-2 model,
+but this method is arguably a little heavyweight.  In this section we
+describe the XUTXO-3 model which provides much more lightweight
+support of NFTs.
+
+The basic idea is to change the \valuemap{} type used in XUTXO-2.  We
+replace it with a type \tokenmap{} which will allow us to support both
+fungible and non-fungible tokens, with no need to change any of the
+other ledger rules: see Figure~\ref{fig:xutxo-3-types}.
+
+\begin{ruledfigure}{H}
+  \begin{align*}
+  \tokenmap &= \Map{\s{CurrencyId}}{\Map{\s{Token}}{\s{Value}}}\\
+  \tokenmap^{\pm} &= \Map{\s{CurrencyId}}{\Map{\s{Token}}{\s{Value}^{\pm}}}\\
+  \end{align*}
+  where
+  \begin{itemize}
+\item \s{CurrencyId} is the type we are using to identify currencies, as in Figure~\ref{fig:xutxo-2-types}.
+\item \s{Token} is a type consisting of identifiers for individual
+  tokens.  These will probably be hashes in practice.
+% \item \s{Quantity} is a type used to measure the amount of a token
+%   that is present, probably the type of natural numbers.  
+  \end{itemize}
+  \caption{Modified \valuemap{} type for XUTXO-3}
+  \label{fig:xutxo-3-types}
+\end{ruledfigure}
+
+\noindent Note that the comparison and addition operations for maps defined in
+Figure~\ref{fig:basic-notation} carry over to the type \tokenmap.
+This allows us to define XUTXO-3 by exactly the same types and
+validity rules as XUTXO-2, simply replacing \valuemap{} by
+\tokenmap{} everywhere.
+
+\medskip \noindent\red{I'm slightly confused about what we want to do here.  In
+  Jann's multicurrency proposal, we have a type \s{Quantity} which is
+  used to measure quantities of tokens, but it's left open-ended
+  exactly what \s{Quantity} should be.  The problem is that if we're
+  following the convention of destroying things by forging
+  negative amounts, then we have to know what ``negative'' means, and
+  I think you need at least an ordered abelian group for that.  Will
+  anyone ever use anything that isn't \Z?}
```

Comment:
I think we can just use Z, anything other than Z would make the ledger rules more complicated (because the core nodes would need to know how to add and compare values of that type)